### PR TITLE
[test] Remove httpclient dependency

### PR DESCRIPTION
**Pull request by dadoonet**
We can provide our own implementation of a simple HTTP client to run all our tests (REST or others) so we don't really need to depend on httpclient and its dependencies.

It will help to avoid also maven dependency issues as reported in https://github.com/elastic/elasticsearch/pull/12036#issuecomment-119591775

We can remove httpclient as well from cloud plugins and let maven find the right transitive version.

Funny stuff:

Cloud-gce is expecting httpclient 4.0.1 but as it was defined in parent pom.xml, version 4.3.6 was used.
Same for azure 4.3.1 vs 4.3.6.

We also add `http.method_override.enabled` parameter:
- `http.method_override.enabled` enables support for `X-HTTP-Method-Override`, `X-HTTP-Method` and `X-METHOD-OVERRIDE` headers.

Defaults to `false`.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java [hunk 1]
```diff
@@ -69,6 +69,14 @@ public NettyHttpRequest(org.jboss.netty.handler.codec.http.HttpRequest request,
...
     public Method method() {
         HttpMethod httpMethod = request.getMethod();
+
+        if (request.headers() != null) {
+            String methodOverride = request.headers().get("X-HTTP-Method-Override");
+            if (methodOverride != null) {
+                httpMethod = HttpMethod.valueOf(methodOverride);
+            }
+        }
+
```
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java]: I want @jaymode to look at this one (see corresponding client usage in HttpClient.java).

Overall I really like this PR, and I like how it abstracts the implementation. But this hack might have implications... if its not ok, the only alternative I can think of right now is to use netty as the implementation?

File: core/src/test/java/org/elasticsearch/test/rest/ElasticsearchRestTestCase.java [hunk 2]
```diff
@@ -295,10 +295,7 @@ private static void validateSpec(RestSpec restSpec) {
 
     @AfterClass
     public static void close() {
-        if (restTestExecutionContext != null) {
-            restTestExecutionContext.close();
-            restTestExecutionContext = null;
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/test/rest/ElasticsearchRestTestCase.java]: there is no need to close connections anymore? The previous http client did some connection pooling too, added here https://github.com/elastic/elasticsearch/commit/7776b6b4f0fc105d9261c1140c683c70e7fc0979 . How does this client work in terms of connections?

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java]: I agree this hack is odd...

**Comment by javanna** [comment]
this looks good but when I look at it I get lost on why we need to remove the http-client dependency. Seems odd that we can't use the library we want in our test framework otherwise some plugins may break, so this seems more like a workaround to me. That said if we want to move to HttpUrlConnection  and remove this dependency regardless I don't have any problem with that.

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java (hunk 1 above)
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java]: Agreed. Actually this PATCH is at least not safe enough.
Someone could send a `GET` request to `/_all` with `X-HTTP-Method-Override: DELETE` and remove everything!

At least, as stated in [RFC 2616 - HTTP 1.1 section 9.1.1](http://www.ietf.org/rfc/rfc2616.txt):

> Implementors should be aware that the software represents the user in
   their interactions over the Internet, and should be careful to allow
   the user to be aware of any actions they might take which may have an
   unexpected significance to themselves or others.
>
>   In particular, the convention has been established that the GET and
   HEAD methods SHOULD NOT have the significance of taking an action
   other than retrieval. These methods ought to be considered "safe".
   This allows user agents to represent other methods, such as POST, PUT
   and DELETE, in a special way, so that the user is made aware of the
   fact that a possibly unsafe action is being requested.

At least I should only accept this if `httpMethod == POST`.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java]: in general it feels really bad that we have to support this in our production code to work around some http client limitations that doesn't allow to send delete (and get?) with body...  the question should be if this makes sense in our production code regardless I think.

Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java]: I agree. I think IMHO that it would make sense but in another PR. I mean this test thing should not shade that we want to support `X-HTTP-Method-Override` (Google) or `X-HTTP-Method` (Microsoft) or `X-METHOD-OVERRIDE` (IBM).

Going to look if we can do something else here in tests. Most likely we can in Java as httpclient solved it already :)

**Comment by rmuir** [comment]
Its not a workaround. You simply cannot pollute your test classpath with dependencies like this, or it will hide bugs.

Thats because if something has a *real* dependency on it, and the classpath is fucked up, nothing will catch you. 

Its just a standard thing with java development, keep the test classpath limited to testing jars. 

Adding commonly used jars like httpclient to the test classpath hides bugs, like cloud plugins not working at all.

File: core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java (hunk 1 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/http/netty/NettyHttpRequest.java]: I'm ok with the updated approach limiting it to `POST` operations. The other thing I would like to see is making this configurable (as in be able to turn it off) and possibly part of the Netty pipeline after the `NettyHttpRequestDecoder`? Maybe off by default is the right choice here and we have to enable it for our tests?

**Comment by dadoonet** [comment]
@jaymode Thanks! I added a new commit to disable by default this feature

**Comment by rjernst** [comment]
Do we really want to make hacks in production code just to fix a potential problem with classpath in tests? I'm all for fixing the issue (Robert is right, we need to not pollute the test classpath with non testy stuff), but can we try the netty route he suggested? I would hate for a user to find this hack, start *using* it on purpose, and then someone make the argument later that we can't remove because it is a "feature".

**Comment by rmuir** [comment]
very simple, don't document it. then it can be removed when we require java 8 or its eliminated via another mechanism.

**Comment by nik9000** [comment]
> very simple, don't document it. then it can be removed when we require java 8 or its eliminated via another mechanism.

That doesn't really work when people can and will read the source code.

**Comment by rmuir** [comment]
It totally works. we add an unsupported api/hack, and then we remove it. Pretty much 100% of what I have added to elasticsearch sits in this category.

If someone builds shit that relies upon that unsupported API, their bad, they will pay the piper. Not my problem.

**Comment by dadoonet** [comment]
Guys? What is the consensus here if any? Should I try to rebase that and merge it in or simply close the PR?

Thanks!

**Comment by clintongormley** [comment]
@dadoonet let's merge this and turn it on only for testing.

thanks

**Comment by dadoonet** [comment]
I rebased the PR so I had to fix some stuff.
For example, I needed to define `System.setProperty("sun.net.http.allowRestrictedHeaders", "true");` in the HTTPClient (test package).

I ran a full `mvn install` and everything looks fine but I'd love to get a final review.

Thanks!

**Comment by dadoonet** [comment]
@rmuir Do you agree with the latest commit I added?

**Comment by dadoonet** [comment]
I'm closing this one. I don't think we will merge it.
Thinking about it more, it actually does not really make sense to forbid the usage of any HttpClient lib in our tests and try to rewrite one from scratch.

We need to solve the dependency issues we might hit in another way IMHO. 

For now, the only "side effect" I see by keeping the current code is with GCE and Azure plugins:

* GCE is using `httpclient 4.0.1`.
* Azure is using `httpclient 4.3.1`.

We defined version `httpclient 4.3.6` and this one is packaged within our GCE/Azure plugins.

I don't see it as an issue for now for Azure. Might be problematic for GCE though.

Related GCE issues:

* https://github.com/google/google-http-java-client/issues/250
* https://github.com/google/google-api-java-client/issues/928

**Comment by rjernst** [comment]
> Thinking about it more, it actually does not really make sense to forbid the usage of any HttpClient lib in our tests and try to rewrite one from scratch.

Why? The reasoning was well laid out. Tests shouldn't be adding non test libs; that just leads to errors down the road. The state of things with different httpclient versions for GCE and Azure is a perfect example of why this needs to happen: we aren't testing the right thing.

**Comment by kimchy** [comment]
I read though this issue and change, I like this change and I think we should pull it in. The less dependencies we have in general, the better, and this change gets us in the right place.

I would put a javadoc on the new setting, saying it will be removed in the future, and not document it. It is the best out of all the options I can think about (having this setting in the first place). We can then remove this setting once we move to Java 8 as a core dependency.

**Comment by jaymode** [comment]
+1 to keeping this going. @dadoonet let me know if you'd like me to help with the truststore usage (looking at the last commit)

**Comment by dadoonet** [comment]
Ok. Reopening.

**Comment by dadoonet** [comment]
@jaymode Yes! I'd love to have your help with the truststore part.
Do you want me to open a public branch in elasticsearch repo so you could push new commits or do you prefer to fork my repo and send PR on it?

**Comment by jaymode** [comment]
@dadoonet sent a PR on your fork, if the change is good with you, the we can just cherry-pick it into your branch.

**Comment by rmuir** [comment]
Can we bring this guy back to life? We require java 8 now, so we should not need the controversial part right? and this would be a huge win for master. It avoids not just dragging in httpclient dependency, but that itself also drags in 2 commons-* dependencies.

**Comment by dadoonet** [comment]
Now that we decided to build an HTTP client:

* does it makes sense to keep this PR?
* if it still makes sense, I think we can probably reduce its surface. Most of the challenges I had to face were for REST tests passing. As REST tests will now use the HTTP client, I think we could reduce the complexity of this PR.

Thoughts?

**Comment by javanna** [comment]
I think we should re-evaluate once the http client is ready and not spend on time on this, especially given that our http client will most likely use the dependency that we were trying to remove here?

**Comment by javanna** [comment]
the new low level RestClient is in, the REST tests infrastructure is using it now. Closing.
